src/rdf4cpp/rdf/Literal.cpp

```diff
@@ -542,81 +609,126 @@ Literal Literal::as_language_tag_eq(Literal const &other, Node::NodeStorage &nod
 }
 
 Literal::operator std::string() const noexcept {
-    // TODO: escape non-standard chars correctly
-    auto const quoted_lexical_into_stream = [](std::ostream &out, std::string_view const lexical) noexcept {
-        // TODO: escape everything that needs to be escaped in N-Tripels/N-Quads
+    if (this->null()) {
+        return "null";
+    }
+
+    auto const escaped_size = [](size_t size) noexcept {
+        return static_cast<size_t>(static_cast<double>(size) * 1.2);
+    };
+
+    auto const append_quoted_lexical = [](std::string &out, std::string_view const lexical) noexcept {
+        // https://www.w3.org/TR/n-triples/#grammar-production-STRING_LITERAL_QUOTE
 
-        out << "\"";
-        for (auto const character : lexical) {
-            switch (character) {
+        out.append("\"");
+        for (char const ch : lexical) {
+            switch (ch) {
+                case '"': {
+                    out.append(R"(\")");
+                    break;
+                }
                 case '\\': {
-                    out << R"(\\)";
+                    out.append(R"(\\)");
                     break;
                 }
                 case '\n': {
-                    out << R"(\n)";
+                    out.append(R"(\n)");
                     break;
                 }
                 case '\r': {
-                    out << R"(\r)";
+                    out.append(R"(\r)");
                     break;
                 }
-                case '"': {
-                    out << R"(\")";
-                    break;
-                }
-                [[likely]] default : {
-                    out << character;
+                [[likely]] default: {
+                    out.push_back(ch);
                     break;
                 }
             }
         }
-        out << "\"";
+        out.append("\"");
     };
 
-    std::ostringstream oss;
+    std::string buf;
 
-    if (this->null()) {
-        oss << "null";
-    }
-    else if (this->is_inlined()) {
-        if (this->datatype_eq<datatypes::rdf::LangString>()) {
-            return static_cast<std::string>(this->lang_tagged_get_de_inlined());
+    if (this->datatype_eq<datatypes::rdf::LangString>()) {
+        auto const value = this->lang_tagged_get_de_inlined().backend_handle().literal_backend().get_lexical();
+        buf.reserve(escaped_size(value.lexical_form.size()) + value.language_tag.size() + 1);
+
+        if (value.needs_escape) [[unlikely]] {
+            append_quoted_lexical(buf, value.lexical_form);
+        } else {
+            buf.push_back('"');
+            buf.append(value.lexical_form);
+            buf.push_back('"');
         }
```

Comment:
reserve buffer in the (value.needs_escape) branches to be more precise in memory usage.